cmd/plugins/memtierd/main.go

```diff
@@ -0,0 +1,520 @@
+	// Name of the QoS class, matches to annotations in
+	// pods. Examples:
+	// annotations:
+	//   # The default for all containers in the pod:
+	//   class.memtierd.nri.io: swap-idle-data
+	//   # Override the default for CONTAINERNAME1:
+	//   class.memtierd.nri.io/CONTAINERNAME1: noswap
+	//   # Do not apply any class for CONTAINERNAME2:
+	//   class.memtierd.nri.io/CONTAINERNAME2: ""
+	Name string
+
+	// MemtierdConfig is a string that contains full configuration
+	// for memtierd. If non-empty, a separate memtierd will be
+	// launched to track each container of this QoS class.
+	MemtierdConfig string
+
+	// AllowSwap: if true, set memory.swap.max to max, if false,
+	// set memory.swap.max to 0. If undefined, do not touch
+	// memory.swap.max. Direct annotation that defines value of
+	// memory.swap.max overrides this option.
+	AllowSwap *bool
+}
+
+type memtierdEnv struct {
+	pid        int
+	ctrDir     string
+	configFile string
+	outputFile string
+	pidFile    string
+	cmd        *exec.Cmd
+}
+
+type options struct {
+	HostRoot string
+}
+
+const (
+	annotationSuffix = ".memtierd.nri.io"
+)
+
+var opt = options{}
+
+var (
+	log *logrus.Logger
+)
+
+// Configure handles connecting to container runtime's NRI server.
+func (p *plugin) Configure(ctx context.Context, config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("Connected to %s %s...", runtime, version)
+	if config != "" {
+		if err := p.setConfig([]byte(config)); err != nil {
+			return 0, loggedErrorf("Configure: loading configuration from NRI server failed: %s", err)
+		}
+		log.Debugf("Using configuration from NRI server")
+	} else {
+		log.Debugf("No configuration from NRI server")
+	}
+	return 0, nil
+}
+
+// setConfig applies new plugin configuration.
+func (p *plugin) setConfig(config []byte) error {
+	log.Tracef("setConfig: parsing\n---8<---\n%s\n--->8---", config)
+	cfg := pluginConfig{}
+	err := yaml.Unmarshal(config, &cfg)
+	if err != nil {
+		log.Tracef("setConfig: parsing failed: %s", err)
+		return fmt.Errorf("setConfig: cannot parse configuration: %w", err)
+	}
+	p.config = &cfg
+	if log.GetLevel() == logrus.TraceLevel {
+		log.Tracef("new configuration has %d classes:", len(p.config.Classes))
+		for _, cls := range p.config.Classes {
+			log.Tracef("- %s", cls.Name)
+		}
+	}
+	return nil
+}
+
+// pprintCtr() returns human readable container name that is
+// unique to the node.
+func pprintCtr(pod *api.PodSandbox, ctr *api.Container) string {
+	return fmt.Sprintf("%s/%s:%s", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+}
+
+// loggedErrorf formats, logs and returns an error.
+func loggedErrorf(s string, args ...any) error {
+	err := fmt.Errorf(s, args...)
+	log.Errorf("%s", err)
+	return err
+}
+
+// associate adds new key-value pair to a map, or updates existing
+// pair if called with the override set. Returns true if the pair was
+// added/updated.
+func associate(m *map[string]string, key, value string, override bool) bool {
```

Comment:
nitpick: why the pointer to map ? Maps are passed by reference and this function doesn't accept nil maps that it would then initialize.

Reply:
Fixed.